iothub/device/src/Authentication/ClientAuthenticationWithSharedAccessKeyRefresh.cs

```diff
@@ -0,0 +1,158 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Net;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Client
+{
+    /// <summary>
+    /// Authentication method that generates shared access signature (SAS) token with refresh, based on a provided shared access key (SAK).
+    /// </summary>
+    public class ClientAuthenticationWithSharedAccessKeyRefresh : ClientAuthenticationWithTokenRefresh
+    {
+        private readonly IotHubConnectionString _iotHubConnectionString;
+
+        /// <summary>
+        /// Creates an instance of this class.
+        /// </summary>
+        /// <param name="sharedAccessKey">Shared access key value.</param>
+        /// <param name="deviceId">Device identifier.</param>
+        /// <param name="moduleId">Module identifier.</param>
+        /// <param name="sharedAccessKeyName">Shared access key name.</param>
+        /// <param name="sasTokenTimeToLive">
+        /// The suggested time to live value for the generated SAS tokens.
+        /// The default value is 1 hour.
+        /// </param>
+        /// <param name="sasTokenRenewalBuffer">
+        /// The time buffer before expiry when the token should be renewed, expressed as a percentage of the time to live.
+        /// The default behavior is that the token will be renewed when it has 15% or less of its lifespan left.
+        /// </param>
+        /// <exception cref="ArgumentNullException">Thrown when <paramref name="sharedAccessKey"/> is null.</exception>
+        /// <exception cref="ArgumentException">
+        /// Thrown when <paramref name="sharedAccessKey"/> or <paramref name="sharedAccessKeyName"/> is empty or whitespace.
+        /// </exception>
+        public ClientAuthenticationWithSharedAccessKeyRefresh(
+            string sharedAccessKey,
+            string deviceId,
+            string moduleId = default,
+            string sharedAccessKeyName = default,
+            TimeSpan sasTokenTimeToLive = default,
+            int sasTokenRenewalBuffer = default)
```

Comment:
The ordering feels a bit off - sharedAccessKey, then device and module Id, and then the shared access key name. 

How would you feel about us creating an overload of this constructor, one with the key name and one without?
We'll have the overhead of an additional constructor, but cleaner ordering of the params.

I don't think this is a blocker of any sort, more of a nit.

Reply:
yeah, I don't like the ordering, either. The only reason doing so is the optional params have to pass in later.

Reply:
Also bring the same changes to ClientAuthenticationWithX509Certificate ([ref](https://github.com/Azure/azure-iot-sdk-csharp/blob/previews/v2/iothub/device/src/Authentication/ClientAuthenticationWithX509Certificate.cs#L29))